wallpaper-play/Service/SystemWallpaperService.swift

```diff
@@ -2,11 +2,43 @@ import Foundation
 import AppKit
 
 protocol SystemWallpaperService {
-    func setWallpaper(_ url: URL) throws
+    //func setWallpaper(_ url: URL) throws
+    func backupWallpapers() throws
+    func restoreWallpaper() throws
 }
 
 class SystemWallpaperServiceImpl: SystemWallpaperService {
     func setWallpaper(_ url: URL) throws {
         try NSWorkspace.shared.setDesktopImageURL(url, for: NSScreen.main!)
     }
+    
+    func backupWallpapers() throws {
+        var wallpapers = [String]()
+        NSScreen.screens.forEach { screen in
+            if let url = NSWorkspace.shared.desktopImageURL(for: screen) {
+                guard url != ApplicationFileManagerImpl().getDirectory(.latestThumb)!.appendingPathComponent("latest.png") else { return } // do not backup the saved thumbnail
+                wallpapers.append(url.path)
+            }
+        }
+        
+        if !wallpapers.isEmpty {
+            try JSONSerialization.data(withJSONObject: wallpapers).write(to: ApplicationFileManagerImpl().getDirectory(.latestThumb)!.appendingPathComponent("wallpaper_backup.json"))
+        }
+        
+    }
+    
+    func restoreWallpaper() throws {
+        let wallpapers = try JSONSerialization.jsonObject(with: Data(contentsOf: ApplicationFileManagerImpl().getDirectory(.latestThumb)!.appendingPathComponent("wallpaper_backup.json")), options: []) as! [String]
+        guard wallpapers.count > 0 else { return }
+        for (idx, screen) in NSScreen.screens.enumerated() {
+            let idx = idx < wallpapers.count-1 ? idx : wallpapers.count-1
+            try NSWorkspace.shared.setDesktopImageURL(URL(fileURLWithPath: wallpapers[idx]), for: screen)
+        }
+        for (idx, wallpaperPath) in wallpapers.enumerated() {
+            if idx < NSScreen.screens.count-1 {
+                try NSWorkspace.shared.setDesktopImageURL(URL(fileURLWithPath: wallpaperPath), for: NSScreen.screens[idx])
+            }
+        }
```

Comment:
I have noticed that the system desktop image does not change when using multiple spaces. For example, if I launch an application on Space 1 and then terminate it on Space 2, the system desktop image on Space 1 ends up being a video thumbnail instead of the original system desktop image. This is because NSScreen.screens only takes physically connected screens into account, so it is not able to change the system desktop images on all spaces.

I need to find a way to modify the system desktop images on all spaces in order to resolve this problem, but I do not know how to do so.

If you know how to do this, could you please tell me?

[This problem demo](https://drive.google.com/file/d/11n8ouS7VMSeXuCDH0j5m3stROoFxOLxr/view?usp=sharing)

Reply:
I think that is he same bug as in System Settings, but now it’s an actual issue. I will try to fix it when I can, but I don’t think it’s possible. There are no other methods to set the wallpaper as `NSWorkspace.shared.setDesktopImageURL()`.

Reply:
Sorry for the late reply.

I am going to release this feature as a beta version because I think some users may have problems with the wallpaper changing without their permission.

Thank you.

Reply:
Ah ok, no problem.

To be honest, I am a bit surprised as well that Apple allows that for apps released in the App Store.

I hope to contribute more to this in the future :).